### PR TITLE
Close #21825: Only lock OpenGL texture cache when using multi-threading

### LINKED ISSUE #21825: Avoid locking the OpenGL TextureCache during single threaded drawing
Profiling showed ~4% of time taken to draw a big zoomed out scene with OpenGL is spent locking and unlocking the texture cache, even when drawing is single threaded. Ideally, there should be no locking in single threaded mode. This is already handled in `TTF.cpp` with `FontLockHelper`.

A PR should apply a similar fix for the the texture cache, ideally refactoring both the texture cache and TTF code to use the same lightweight object.  `gConfigGeneral.MultiThreading` should probably also be made atomic.

### PR DESCRIPTION
**Pull request by mrmbernardi**
Fixes #21825 by making  'DrawingLocks' which lock only if multi threaded drawing is enabled. Refactors TTF.cpp to use this.

Also cleans up some header files, modernises constants, and removes code to avoid using `shared_mutex` on MacOS builds. This was originally added because older versions of MacOS didn't support this in the STL.

### CODE REVIEW

File: src/openrct2/config/Config.cpp [hunk 1]
```diff
@@ -20,7 +20,6 @@
 #include "../drawing/IDrawingEngine.h"
 #include "../interface/Window.h"
 #include "../localisation/Currency.h"
-#include "../localisation/Date.h"
```
[the remark this answers is not in the document]
Comment (mrmbernardi) [reply src/openrct2/config/Config.cpp:23]: I figured it was appropriate to clean up the parts of the code that you work on. That's why I also included the changes to the constant values. I can remove this change if it's a sticking point.

File: src/openrct2/config/Config.h [hunk 2]
```diff
@@ -42,7 +42,7 @@ struct GeneralConfiguration
     bool UncapFPS;
     bool UseVSync;
     bool ShowFPS;
-    bool MultiThreading;
+    std::atomic_uint8_t MultiThreading;
```
Comment (ZehMatt) [inline src/openrct2/config/Config.h:45]: Why not just
```suggestion
    std::atomic<bool> MultiThreading;
```
?

Comment (mrmbernardi) [reply src/openrct2/config/Config.h:45]: Bool doesn't support the toggle used here:

https://github.com/OpenRCT2/OpenRCT2/blob/339a572d03df5f0bfe5a71982c26f30606bbd762/src/openrct2-ui/windows/Options.cpp#L707

Comment (ZehMatt) [reply src/openrct2/config/Config.h:45]: Would it work if you replaced 1 with true?

Comment (mrmbernardi) [reply src/openrct2/config/Config.h:45]: I tried. The issue is that there's no atomic implementation of ^= for bool, for whatever reason. There is such an implementation for integers however.

Ultimately, using atomic here probably isn't really necessary as reading and writing are separated quite a lot in the code, and they have to be for things to work anyway as the value gets saved in the lock type I introduced. But it is a 'best practice' thing as far as I know.

Comment (mrmbernardi) [reply src/openrct2/config/Config.h:45]: This SO post talks about it and it was also modified quite recently:
https://stackoverflow.com/a/9806613

Comment (ZehMatt) [reply src/openrct2/config/Config.h:45]: Hm I see, it probably doesn't need to be atomic, its not going to change the value while rendering.

File: src/openrct2/config/Config.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (janisozaur) [reply src/openrct2/config/Config.cpp:23]: Not a deal breaker for this, but it certainly makes reviewing harder - you could at least split it in two commits

File: src/openrct2/drawing/DrawingLock.hpp [hunk 3]
```diff
@@ -0,0 +1,52 @@
...
+    {
+        if (_enabled)
+            _mutex.lock_shared();
+    }
+    ~DrawingSharedLock()
+    {
+        if (_enabled)
+            _mutex.unlock_shared();
+    }
+};
```
Comment (janisozaur) [inline src/openrct2/drawing/DrawingLock.hpp:52]:
```suggestion
};

```

**Review by janisozaur — APPROVED** [review]

File: src/openrct2/config/Config.cpp (hunk 1 above)
[the remark this answers is not in the document]
Comment (mrmbernardi) [reply src/openrct2/config/Config.cpp:23]: Thanks, I'll keep that in mind for future PRs.

**Pull request merged**